core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -169,70 +284,116 @@ public void segmentsLost(IntSet lostSegments) {
-               .map(segment -> {
-                  Flowable<I> innerFlowable = Flowable.fromPublisher(set.localPublisher(segment))
-                        // If we complete the iteration try to remove the segment - so it can't be suspected
-                        .doOnComplete(() -> concurrentSegments.remove(segment));
-                  if (keysToExclude != null) {
-                     innerFlowable = innerFlowable.filter(i -> !keysToExclude.contains(toKeyFunction.apply(i)));
-                  }
+         resultFlowable = exactlyOnceParallel(set, keysToExclude, toKeyFunction, segments, transformer,
+               listener, concurrentSegments);
+      } else {
+         resultFlowable = exactlyOnceSequential(set, keysToExclude, toKeyFunction, segments, transformer,
+               listener, concurrentSegments);
+      }
+      return exactlyOnceHandleLostSegments(finalizer.apply(resultFlowable), listener);
+   }
 
-                  CompletionStage<R> stage = transformer.apply(innerFlowable);
-                  CompletableFuture<R> future = stage.toCompletableFuture();
-                  // This will always be true unless there is a store
-                  if (future.isDone()) {
-                     if (listener.segmentsLost.contains(segment)) {
-                        return CompletableFutures.<R>completedNull();
-                     }
-                     return future;
-                  }
+   protected <R> CompletionStage<PublisherResult<R>> exactlyOnceHandleLostSegments(CompletionStage<R> finalValue, SegmentListener listener) {
+      return handleLostSegments(finalValue, listener);
+   }
 
-                  return stage.thenCompose(value -> {
-                     // This means the segment was lost in the middle of processing
-                     if (listener.segmentsLost.contains(segment)) {
-                        return CompletableFutures.completedNull();
-                     }
-                     return CompletableFuture.completedFuture(value);
-                  });
-               }).sequential();
+   /**
+    * This method iteratively submits a task to operate on the cpu bound thread pool for each segment that is available.
+    * The tasks perform a type of work stealing where they attempt to retrieve the next available segment and process
+    * them as fast as possible. It is entirely possible that a given task is never submitted due to the other tasks
+    * completing all the segments asynchronously. After the main thread has submitted all the tasks it will attempt
+    * to steal a segment and run it if possible and if it can will subsequently attempt to complete all remaining
+    * segments in the same fashion as the other threads. NOTE that this behavior is not normally how reactive
+    * streams are done as given operations are not normally performed until the returned Flowable is subscribed to, but
+    * for performance reasons this method eagerly publishes entries.
```

Comment:
> Can you detail the performance reasons?

Sure, I will add it. But it is because we use the invoking thread to limit context switching. Normally you would submit all the tasks to the other scheduler to perform the work, but since we control the entire Flowable lifecycle we can optimize to do this.

> And is the "operations are not normally performed until the returned Flowable is subscribed to" part of the reactive streams spec, or is it ok either way, only most streams choose to do it one way?

I don't believe it is part of the spec, but heavily recommended as they want to do everything as lazily as possible. The reason this is done is actually because of the answer to your first question.